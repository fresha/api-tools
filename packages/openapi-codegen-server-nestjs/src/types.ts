import type { DTO } from './DTO';
import type { Nullable } from '@fresha/api-tools-core';
import type { Context as BasicContext } from '@fresha/openapi-codegen-utils';
import type { SchemaModel } from '@fresha/openapi-model/build/3.0.3';
import type { Project } from 'ts-morph';

export interface Context extends BasicContext {
  readonly project: Project;
  readonly nestApp: string;

  addDTO(name: string, schema: Nullable<SchemaModel>): DTO;
}
