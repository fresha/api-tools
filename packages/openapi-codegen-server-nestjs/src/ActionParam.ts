import assert from 'assert';

import { addDecorator, addImportDeclaration, Logger } from '@fresha/openapi-codegen-utils';

import type { Action } from './Action';
import type { Nullable } from '@fresha/api-tools-core';
import type { SchemaModel } from '@fresha/openapi-model/build/3.0.3';
import type { MethodDeclaration } from 'ts-morph';

/**
 * Generates code for a single parameter of a controller action.
 */
export class ActionParam {
  readonly action: Action;
  readonly from: 'body' | 'path' | 'query'; // parameter source
  readonly name: string; // parameter (method argument) name
  private readonly schema: Nullable<SchemaModel>; // OpenAPI schema
  protected readonly logger: Logger;

  constructor(
    action: Action,
    param: {
      in: 'path' | 'query' | 'body' | 'cookie' | 'header';
      name: string;
      schema: Nullable<SchemaModel>;
    },
    logger: Logger,
  ) {
    assert(param.in === 'path' || param.in === 'query' || param.in === 'body');

    this.action = action;
    this.from = param.in;
    this.name = param.name;
    this.schema = param.schema;
    this.logger = logger;
  }

  generateCode(methodDecl: MethodDeclaration): void {
    const sourceFile = methodDecl.getSourceFile();

    let decoratorName = '';
    switch (this.from) {
      case 'body':
        decoratorName = 'Body';
        break;
      case 'path':
        decoratorName = 'Param';
        break;
      case 'query':
        decoratorName = 'Query';
        break;
      default:
        assert.fail(`Unsupported parameter source ${String(this.from)}`);
    }

    addImportDeclaration(sourceFile, '@nestjs/common', decoratorName);

    let paramType = 'string';
    let validatorName: string | undefined;

    if (this.schema) {
      switch (this.schema?.type) {
        case null:
          break;
        case 'boolean':
          paramType = 'boolean';
          validatorName = 'ParseBoolPipe';
          break;
        case 'integer':
          paramType = 'number';
          validatorName = 'ParseIntPipe';
          break;
        case 'number':
          paramType = 'number';
          validatorName = 'ParseFloatPipe';
          break;
        case 'string':
          break;
        default:
          if (this.from !== 'body') {
            assert.fail(`Unsupported schema type ${this.schema.type}`);
          }
      }
    }

    if (validatorName) {
      addImportDeclaration(methodDecl.getSourceFile(), '@nestjs/common', validatorName);
    }

    const actionParam = methodDecl.addParameter({
      name: this.name,
      type: paramType,
    });

    switch (this.from) {
      case 'body':
        addDecorator(actionParam, 'Body', this.name);
        break;
      default: {
        const decorator = addDecorator(actionParam, 'Param', this.name);
        if (validatorName) {
          decorator.addArgument(validatorName);
        }
        break;
      }
    }
  }
}
