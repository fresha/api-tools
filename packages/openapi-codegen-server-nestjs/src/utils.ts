import assert from 'assert';
import fs from 'fs';
import path from 'path';

type NestCliJSON = {
  sourceRoot: string;
  projects?: Record<
    string,
    {
      sourceRoot: string;
    }
  >;
};

/**
 * For given directory of a NestJS application, and name of a sub-app therein, returns
 * the root folder for that sub-app. The information is read from nest-cli.json file.
 */
export const getNestJSSubAppPath = (packageDir: string, appName?: string): string => {
  const filePath = path.join(packageDir, 'nest-cli.json');
  assert(fs.existsSync(filePath), `${packageDir} does not seem to be a NestJS application`);

  const json = JSON.parse(fs.readFileSync(filePath, 'utf-8')) as NestCliJSON;

  if (appName) {
    assert(json.projects);
    const result = json.projects[appName]?.sourceRoot;
    assert(
      !!result,
      `Cannot determine source root of "${appName}" app in ${packageDir}/nest-cli.json`,
    );

    return path.join(packageDir, result);
  }

  const result = json.sourceRoot;
  assert(!!result, `Cannot determine source root in ${packageDir}/nest-cli.json`);

  return path.join(packageDir, result);
};
