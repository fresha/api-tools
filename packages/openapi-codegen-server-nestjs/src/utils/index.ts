export * from './lodash';
export * from './nest';
export * from './openapi';
export * from './ts';
export * from './utils';
