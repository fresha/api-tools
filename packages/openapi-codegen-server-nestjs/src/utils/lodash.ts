export { default as camelCase } from 'lodash.camelcase';
export { default as kebabCase } from 'lodash.kebabcase';
export { default as startCase } from 'lodash.startcase';
