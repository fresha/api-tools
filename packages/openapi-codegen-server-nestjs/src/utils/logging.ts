import * as winston from 'winston';

export type Logger = winston.Logger;

export const createLogger = (verbose: boolean): Logger => {
  return winston.createLogger({
    level: verbose ? 'info' : 'warn',
    transports: [new winston.transports.Console({ format: winston.format.simple() })],
  });
};
