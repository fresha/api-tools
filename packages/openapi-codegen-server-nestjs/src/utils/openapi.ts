export const significantNameParts = (pathUrl: string): string[] =>
  pathUrl.split('/').filter(x => x && !(x.startsWith('{') && x.endsWith('}')));
