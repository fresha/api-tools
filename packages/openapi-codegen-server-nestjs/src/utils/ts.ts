import {
  SourceFile,
  ts,
  printNode,
  DecoratableNode,
  Node,
  ImportSpecifier,
  Decorator,
} from 'ts-morph';

export const addNamedImport = (
  sourceFile: SourceFile,
  moduleSpecifier: string,
  importedName: string,
  localName?: string,
): ImportSpecifier | undefined => {
  let importDecl = sourceFile.getImportDeclaration(
    decl => decl.getModuleSpecifier().getLiteralValue() === moduleSpecifier,
  );
  if (!importDecl) {
    importDecl = sourceFile.addImportDeclaration({ moduleSpecifier });
  }

  const namedImports = importDecl.getNamedImports();
  if (!namedImports.find(elem => elem.getName() === importedName)) {
    return importDecl.addNamedImport({ name: importedName, alias: localName });
  }

  return undefined;
};

export const addCommonNestImports = (node: Node, ...importedNames: string[]): void => {
  const sourceFile = node.getSourceFile();
  for (const name of importedNames) {
    addNamedImport(sourceFile, '@nestjs/common', name);
  }
};

export const addClassValidatorImports = (node: Node, ...importedNames: string[]): void => {
  const sourceFile = node.getSourceFile();
  for (const name of importedNames) {
    addNamedImport(sourceFile, 'class-validator', name);
  }
};

export const addDecorator = (
  nodeToAdd: DecoratableNode,
  name: string,
  ...args: (string | number | boolean | undefined)[]
): Decorator => {
  const decorator = nodeToAdd.addDecorator({ name });
  if (args.length) {
    decorator.setIsDecoratorFactory(true);
  }
  for (const arg of args) {
    if (arg !== undefined) {
      if (arg === true) {
        decorator.addArgument(printNode(ts.factory.createTrue()));
      } else if (arg === false) {
        decorator.addArgument(printNode(ts.factory.createFalse()));
      } else if (typeof arg === 'string') {
        decorator.addArgument(printNode(ts.factory.createStringLiteral(arg, true)));
      } else if (typeof arg === 'number') {
        decorator.addArgument(printNode(ts.factory.createNumericLiteral(arg)));
      }
    }
  }
  return decorator;
};
