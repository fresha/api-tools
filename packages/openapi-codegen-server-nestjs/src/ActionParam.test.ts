import { OpenAPIFactory, OpenAPIModel } from '@fresha/openapi-model/build/3.0.3';
import { MethodDeclaration, Project } from 'ts-morph';

import '@fresha/jest-config/build/types';

import { ActionParam } from './ActionParam';
import { createLogger } from './utils/logging';

import type { Action } from './Action';

let openapi: OpenAPIModel = {} as OpenAPIModel;
let methodDecl: MethodDeclaration = {} as MethodDeclaration;

const logger = createLogger(false);

beforeEach(() => {
  openapi = OpenAPIFactory.create();
  openapi.components.setSchema('BooleanParam', 'boolean');
  openapi.components.setSchema('ObjectParam', 'object');

  const tsProject = new Project({ useInMemoryFileSystem: true });
  const tsSourcFile = tsProject.createSourceFile(
    'param.ts',
    `class X {
      y() {}
     }`,
  );
  methodDecl = tsSourcFile.getClassOrThrow('X').getMethodOrThrow('y');
});

test('construction', () => {
  const param = new ActionParam({} as Action, { in: 'path', name: 'id', schema: null }, logger);
  param.generateCode(methodDecl);

  expect(methodDecl.getSourceFile()).toHaveFormattedText(
    `import { Param } from '@nestjs/common';

    class X {
      y(@Param('id') id: string) {}
    }`,
  );
});

test('typed parameters', () => {
  const param = new ActionParam(
    {} as Action,
    {
      in: 'path',
      name: 'id',
      schema: openapi.components.schemas.get('BooleanParam')!,
    },
    logger,
  );
  param.generateCode(methodDecl);

  expect(methodDecl.getSourceFile()).toHaveFormattedText(
    `import { Param, ParseBoolPipe } from '@nestjs/common';

    class X {
      y(@Param('id', ParseBoolPipe) id: boolean) {}
    }`,
  );
});

test('does not support non-primitive parameter schema', () => {
  const param = new ActionParam(
    {} as Action,
    {
      in: 'path',
      name: 'obj',
      schema: openapi.components.schemas.get('ObjectParam')!,
    },
    logger,
  );

  expect(() => param.generateCode(methodDecl)).toThrow();
});
