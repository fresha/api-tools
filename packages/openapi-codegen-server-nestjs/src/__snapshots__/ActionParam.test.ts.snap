// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`construction 1`] = `
"import { Param } from "@nestjs/common";

class X {
      y(@Param('id') id: string) {}
     }"
`;

exports[`typed parameters 1`] = `
"import { Param, ParseBoolPipe } from "@nestjs/common";

class X {
      y(@Param('id', ParseBoolPipe) id: boolean) {}
     }"
`;
